File: 008/StringToInteger.java
public class StringToInteger {
    public int myAtoi(String str) {
        int i = 0, ans = 0, sign = 1, len = str.length();
        while (i < len && str.charAt(i) == ' ')
            ++i;
        if (i < len && (str.charAt(i) == '-' || str.charAt(i) == '+')) {
            sign = str.charAt(i++) == '+' ? 1 : -1;
        }
        for (; i < len; ++i) {
            int tmp = str.charAt(i) - '0';
            if (tmp < 0 || tmp > 9)
                break;
            if (ans > Integer.MAX_VALUE / 10 || ans == Integer.MAX_VALUE / 10 && Integer.MAX_VALUE % 10 < tmp)
                return sign == 1 ? Integer.MAX_VALUE : Integer.MIN_VALUE;
            else
                ans = ans * 10 + tmp;
        }
        return sign * ans;
    }
}

File: 008/StringToInteger.kt
class StringToInteger {

    fun myAtoi(str: String): Int {
        val s = str.trim()
        if (s.isEmpty()) {
            return 0
        }
        if (s.length == 1) {
            return if (s[0] <= '0' || s[0] >= '9') 0 else s.toInt()
        }

        val plus = s[0] == '+'
        val minus = s[0] == '-'

        var result = 0
        val startIndex = if (plus || minus) 1 else 0
        for (i in startIndex until s.length) {
            if (s[i] in '0'..'9') {
                if (Integer.MAX_VALUE / 10 - (s[i] - '0') <= result) {
                    if (minus && result * 10 + (s[i] - '0') == Integer.MAX_VALUE) {
                        return -Integer.MAX_VALUE
                    }
                    return if (minus) Integer.MIN_VALUE else Integer.MAX_VALUE
                }
                result = result * 10 + (s[i] - '0')
            } else {
                return if (minus) -result else result
            }
        }

        return if (minus) -result else result
    }

}